quickwit-storage/src/storage_with_upload_cache.rs

```diff
@@ -0,0 +1,885 @@
+    pub items: Vec<(PathBuf, usize)>,
+}
+
+impl CacheState {
+    /// Construct an instance of [`CacheState`] from a persisted cache state file.
+    pub fn from_path(path: &Path) -> StorageResult<Self> {
+        let file_path = path.to_path_buf().join(CACHE_STATE_FILE_NAME);
+        let json_file = std::fs::File::open(file_path)?;
+        let reader = std::io::BufReader::new(json_file);
+        serde_json::from_reader(reader)
+            .map_err(|err| StorageErrorKind::InternalError.with_error(err))
+    }
+}
+
+/// A struct to wrap the lru cache and its info.
+pub struct WrappedCache {
+    /// Underlying L.R.U cache.
+    pub lru: LruCache<PathBuf, usize>,
+    /// Current number of bytes in the cache.
+    pub num_bytes: usize,
+    /// Current number of files in the cache.
+    pub num_files: usize,
+}
+
+/// A storage with a backing file system cache.
+pub struct StorageWithUploadCache {
+    /// The remote storage.
+    remote_storage: Arc<dyn Storage>,
+    /// The backing local storage.
+    local_storage: Arc<dyn Storage>,
+    /// The localstorage root.
+    local_storage_root: PathBuf,
+    /// The capacity of the cache.
+    disk_capacity: DiskCapacity,
+    /// The underlying cache.
+    disk_cache: Mutex<WrappedCache>,
+}
+
+impl StorageWithUploadCache {
+    /// Create an instance of [`StorageWithUploadCache`]
+    ///
+    /// It needs both the remote and local storage to work with.
+    /// max_item_count and max_num_bytes are used for the cache [`Capacity`].
+    pub fn create(
+        remote_storage: Arc<dyn Storage>,
+        local_storage: Arc<dyn Storage>,
+        max_num_files: usize,
+        max_num_bytes: usize,
+    ) -> StorageResult<Self> {
+        // Extract the local storage root path, also validate that local storage
+        // is of type [`LocalFileStorage]
+        let local_storage_root_path =
+            LocalFileStorage::extract_root_path_from_uri(&local_storage.uri())?;
+
+        Ok(Self {
+            remote_storage,
+            local_storage,
+            local_storage_root: local_storage_root_path,
+            disk_capacity: DiskCapacity {
+                max_num_files,
+                max_num_bytes,
+            },
+            disk_cache: Mutex::new(WrappedCache {
+                lru: LruCache::unbounded(),
+                num_bytes: 0,
+                num_files: 0,
+            }),
+        })
+    }
+
+    /// Create an instance of [`StorageWithUploadCache`] from a path.
+    ///
+    /// It needs a folder `path` previously used by an instance of [`StorageWithUploadCache`].
+    /// Lastly, the file at `{path}/[`CACHE_STATE_FILE_NAME`]` should be present with valid content.
+    pub fn from_path(
+        remote_storage: Arc<dyn Storage>,
+        storage_uri_resolver: &StorageUriResolver,
+        path: &Path,
+    ) -> StorageResult<Self> {
+        let cache_state = CacheState::from_path(path)?;
+        let local_storage = storage_uri_resolver
```

Comment:
It feels strange to store the `local_storage_uri` in the cache state.
Why not remove `local_storage_uri` from the cache state and just use the path? 
By the way, renaming `path` by `cache_dir` would make things clearer.

Can we say that `local_storage` is the `cache_storage`?

Reply:
> Can we say that local_storage is the cache_storage?
yes

Reply:
I don't think we want a `local_storage` at all for the local caching directory.